File: main/the-quickest-way-up/the-quickest-way-up.cs
using System;
using System.Collections.Generic;
using System.Linq;

internal static class Solution {
    private const int NoSolution = -1, Visited = 0, End = 100;
    
    private static int ReadValue() => int.Parse(Console.ReadLine());
    
    private static void ReadConnections(int[] board)
    {
        for (var n = ReadValue(); n > 0; --n) {
            var a = Array.ConvertAll(Console.ReadLine().Split(), int.Parse);
            board[a[0]] = a[1];
        }
    }
    
    private static int[] ReadBoard()
    {
        var board = Enumerable.Range(0, End + 1).ToArray();
        
        ReadConnections(board); // ladders
        ReadConnections(board); // snakes
        
        return board;
    }
    
    private static int BreadthFirstSearch(this int[] board)
    {
        var q = new Queue<int>();
        q.Enqueue(1);
        board[1] = Visited;
        
        for (var turns = 1; q.Count != 0; ++turns) {
            for (var len = q.Count; len != 0; --len) {
                var i = q.Dequeue();
                
                for (var j = Math.Min(End, i + 6); j != i; --j) {
                    switch (board[j]) {
                    case Visited:
                        continue;
                        
                    case End:
                        return turns;
                        
                    default:
                        q.Enqueue(board[j]);
                        board[j] = Visited;
                        continue;
                    }
                }
            }
        }
        
        return NoSolution;
    }
    
    private static void Main()
    {
        for (var t = ReadValue(); t > 0; --t)
            Console.WriteLine(ReadBoard().BreadthFirstSearch());
    }
}

File: main/the-quickest-way-up/the-quickest-way-up.cpp
#include <algorithm>
#include <array>
#include <cstddef>
#include <iostream>
#include <iterator>
#include <numeric>
#include <queue>
#include <stdexcept>

namespace {
    constexpr std::size_t visited {0u}, ending {100u}, max_roll {6u};

    using Board = std::array<std::size_t, ending + 1u>;

    inline void validate(const std::size_t pos)
    {
        if (pos == 0u || pos > ending)
            throw std::out_of_range{"position outside board"};
    }

    void read_connections(Board& board)
    {
        int n = 0;
        for (std::cin >> n; n > 0; --n) {
            std::size_t src {}, dest {};
            std::cin >> src >> dest;
            
            validate(src);
            validate(dest);

            board[src] = dest;
        }
    }

    Board read_board()
    {
        Board board;
        std::iota(std::begin(board), std::end(board), 0u);

        read_connections(board); // ladders
        read_connections(board); // snakes

        return board;
    }

    inline void enqueue_reach(std::queue<std::size_t>& q,
                              const std::size_t src)
    {
        for (auto dest = std::min(ending, src + max_roll); dest != src; --dest)
            q.push(dest);
    }

    int bfs(Board& board)
    {
        std::queue<std::size_t> q;
        q.push(1u);

        for (auto turns = 0; !q.empty(); ++turns) {
            for (auto len = q.size(); len != 0u; --len) {
                const auto pos = board[q.front()];
                q.pop();
                
                switch (board[pos]) {
                case visited:
                    continue;

                case ending:
                    return turns;

                default:
                    board[pos] = visited;
                    enqueue_reach(q, pos);
                }
            }
        }

        return -1;
    }
}

int main()
{
    auto t = 0;
    for (std::cin >> t; t > 0; --t) {
        auto board = read_board();
        std::cout << bfs(board) << '\n';
    }
}
